### PR TITLE
Auto convert to contiguous format for all_gather

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/4781
When syncing using `gather_all`, tensors are required to be in contiguous memory format. This PR converts tensors that are not and warns the user. 

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/metrics/test_ddp.py [hunk 1]
```diff
@@ -43,3 +44,34 @@ def _test_ddp_sum_cat(rank, worldsize):
...
+    with pytest.warns(
+        UserWarning,
+        match="Syncing with `gather_all` requires input to be contiguous"
+        " memory allocated. Will convert to contiguous format.",
+    ):
+        val = metric.compute()
+
+
+@pytest.mark.skipif(sys.platform == "win32", reason="DDP not available on windows")
+def test_not_contiguous_memory_warning():
```
Comment (carmocca) [inline tests/metrics/test_ddp.py:76]: You can add a new case to the `test_ddp` test right above:

```python
pytest.param(_test_memory_warning, marks=pytest.mark.skipif(sys.platform == "win32", reason="DDP not available on windows"))
```
([docs](https://docs.pytest.org/en/latest/skipping.html#skip-xfail-with-parametrize))

File: tests/metrics/test_ddp.py [hunk 2]
```diff
@@ -43,3 +44,34 @@ def _test_ddp_sum_cat(rank, worldsize):
...
+            return x.sum()
+
+    metric = DummyMetric()
+    metric.update(torch.randn(10, 5)[:, 0])
+
+    with pytest.warns(
+        UserWarning,
+        match="Syncing with `gather_all` requires input to be contiguous"
+        " memory allocated. Will convert to contiguous format.",
+    ):
```
Comment (Borda) [inline tests/metrics/test_ddp.py:71]:
```suggestion
    with pytest.warns(UserWarning, match="Syncing with `gather_all` requires input to be contiguous"):
```

File: tests/metrics/test_ddp.py (hunk 1 above)
Comment (Borda) [reply tests/metrics/test_ddp.py:76]: I would keep the pattern the same as everywhere...

**Review by Borda — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/utilities/distributed.py [hunk 3]
```diff
@@ -89,6 +89,12 @@ def gather_all_tensors(result: Union[torch.Tensor], group: Optional[Any] = None)
     if group is None:
         group = torch.distributed.group.WORLD
 
+    if not result.is_contiguous():
+        rank_zero_warn('Syncing with `gather_all` requires input to be contiguous'
+                       ' memory allocated. Will convert to contiguous format.',
+                       UserWarning)
+        result = result.contiguous()
```
Comment (justusschock) [inline pytorch_lightning/utilities/distributed.py:96]:
```suggestion
    result = result.contiguous()
```

The docs on `torch.Tensor.contiguous` state:
> If self tensor is already in the specified memory format, this function returns the self tensor. 

This means, that it should be a no-op if tensor is already contiguous and thus we don't need the check here.

File: tests/metrics/test_ddp.py [hunk 4]
```diff
@@ -43,3 +44,30 @@ def _test_ddp_sum_cat(rank, worldsize):
...
+            return x.sum()
+
+    metric = DummyMetric()
+    metric.update(torch.randn(10, 5)[:, 0])
+
+    with pytest.warns(UserWarning, match="Syncing with `gather_all` requires input to be contiguous"):
+        val = metric.compute()
+
+
+@pytest.mark.skipif(sys.platform == "win32", reason="DDP not available on windows")
```
Comment (justusschock) [inline tests/metrics/test_ddp.py:71]: of course if we remove the condition and the warning above, this test is no longer valid, instead we could check it to just run.

**Review by justusschock — APPROVED** [review]

**Comment by Borda** [comment]
@SkafteNicki mind check the latest comments? 🐰

File: pytorch_lightning/utilities/distributed.py (hunk 3 above)
Comment (SkafteNicki) [reply pytorch_lightning/utilities/distributed.py]: thanks, learned something new. Changed to your suggestion 👍

**Pull request merged**